src/WPGraphQL.php

```diff
@@ -422,50 +424,115 @@ public function init_admin() {
+	 */
+	public static function setup_default_taxonomies( $args, $taxonomy ) {
 		// Adds GraphQL support for categories.
-		if ( isset( $wp_taxonomies['category'] ) ) {
-			$wp_taxonomies['category']->show_in_graphql     = true;
-			$wp_taxonomies['category']->graphql_single_name = 'category';
-			$wp_taxonomies['category']->graphql_plural_name = 'categories';
+		if ( 'category' === $taxonomy ) {
+			$args['show_in_graphql']     = true;
+			$args['graphql_single_name'] = 'category';
+			$args['graphql_plural_name'] = 'categories';
+		} elseif ( 'post_tag' === $taxonomy ) { // Adds GraphQL support for tags.
+			$args['show_in_graphql']     = true;
+			$args['graphql_single_name'] = 'tag';
+			$args['graphql_plural_name'] = 'tags';
+		} elseif ( 'post_format' === $taxonomy ) { // Adds GraphQL support for post formats.
+			$args['show_in_graphql']     = true;
+			$args['graphql_single_name'] = 'postFormat';
+			$args['graphql_plural_name'] = 'postFormats';
 		}
 
-		// Adds GraphQL support for tags.
-		if ( isset( $wp_taxonomies['post_tag'] ) ) {
-			$wp_taxonomies['post_tag']->show_in_graphql     = true;
-			$wp_taxonomies['post_tag']->graphql_single_name = 'tag';
-			$wp_taxonomies['post_tag']->graphql_plural_name = 'tags';
+		return $args;
+	}
+
+	/**
+	 * This sets the post type /taxonomy GraphQL properties.
+	 *
+	 * @param array  $args      Array of arguments for registering the type.
+	 * @param string $wp_type   Post type / Taxonomy type key.
+	 *
+	 * @return array
+	 */
+	public static function set_wp_type_args( $args, $wp_type ) {
+
+		// Bail early if the post type is hidden from the WPGraphQL schema.
+		if ( empty( $args['show_in_graphql'] ) ) {
+			return $args;
 		}
 
-		// Adds GraphQL support for post formats.
-		if ( isset( $wp_taxonomies['post_format'] ) ) {
-			$wp_taxonomies['post_format']->show_in_graphql     = true;
-			$wp_taxonomies['post_format']->graphql_single_name = 'postFormat';
-			$wp_taxonomies['post_format']->graphql_plural_name = 'postFormats';
+		$defaults = [
+			// The "kind" of GraphQL type to register. Can be `interface`, `object`, or `union`.
+			'graphql_kind'                     => 'object',
+			// The callback used to resolve the type. Only used if `graphql_kind` is an `interface` or `union`.
+			'graphql_resolve_type'             => null,
+			// An array of custom interfaces the type should implement.
+			'graphql_interfaces'               => [],
+			// An array of default interfaces the type should exclude.
+			'graphql_exclude_interfaces'       => [],
+			// An array of custom connections the type should implement.
+			'graphql_connections'              => [],
+			// An array of default connection field names the type should exclude.
+			'graphql_exclude_connections'      => [],
+			// Whether to register default connections to the schema.
+			'graphql_register_root_field'      => true,
+			'graphql_register_root_connection' => true,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
```

Comment:
@jasonbahl I think this would be the perfect place for an `apply_filters()`, so users can edit the post type / taxonomy args without worrying about the priority of their `register_{post_type|taxonomy}_args` hooks, but i'm not sure what to call it.

We only have the cpt/tax _name_, not whether the item is a cpt/tax, and `grapqhl_register_{my_custom_wp_type}_args` seems ambiguous to what `args` are actually being filtered.

We could wrap the whole `set_wp_type_args()` in a callable that passes a `string $wp_type post_type|taxonomy` depending on which WP hook is called, but that feels messy to me.

Reply:
> but i'm not sure what to call it.

I'm thinking we can change the filters to:

```php
add_filter( 'register_post_type_args', [ __CLASS__, 'register_graphql_post_type_args' ], 99, 2 );
add_filter( 'register_taxonomy_args', [ __CLASS__, 'register_graphql_taxonomy_args' ], 99, 2 );
```

Then, have 2 functions like so: 

```php
	/**
	 * Set the GraphQL Post Type Args and pass them through a filter.
	 *
	 * @param array  $args           The graphql specific args for the post type
	 * @param string $post_type_name The name of the post type being registered
	 *
	 * @return array
	 * @throws Exception
	 */
	public static function register_graphql_post_type_args( array $args, string $post_type_name ) {

		$args = self::set_wp_type_args( $args, $post_type_name );

		/**
		 * Filters the graphql args set on a post type
		 *
		 * @param array  $args           The graphql specific args for the post type
		 * @param string $post_type_name The name of the post type being registered
		 */
		return apply_filters( 'register_graphql_post_type_args', $args, $post_type_name );

	}

	/**
	 * Set the GraphQL Taxonomy Args and pass them through a filter.
	 *
	 * @param array  $args          The graphql specific args for the taxonomy
	 * @param string $taxonomy_name The name of the taxonomy being registered
	 *
	 * @return array
	 * @throws Exception
	 */
	public static function register_graphql_taxonomy_args( array $args, string $taxonomy_name ) {

		$args = self::set_wp_type_args( $args, $taxonomy_name );

		/**
		 * Filters the graphql args set on a taxonomy
		 *
		 * @param array  $args          The graphql specific args for the taxonomy
		 * @param string $taxonomy_name The name of the taxonomy being registered
		 */
		return apply_filters( 'register_graphql_taxonomy_args', $args, $taxonomy_name );

	}
```

Now, similar to how there is `register_taxonomy_args` and `register_post_type_args` in core, there can be `register_graphql_taxonomy_args` and `register_graphql_post_type_args`

Reply:
My snippets might not be fully comprehensive, but probably close-ish.